Allow reuse and cross-referencing of schemas

risicle opened the pull request:
~~(Sits on top of #4)~~ rebased

https://trello.com/c/PgK7v6dk

See how we can now easily refer to other schemas with e.g. `"$ref": "package_base"` or keep commonly reused patterns like `"$ref": "common#/definitions/slug"`. It also allows us to factor all common response wrappers into an `action_response_base` schema.

risicle committed:
schemas: set these up to be reusable and cross-reference-able

split out organization and package representations in separate files

risicle committed:
schemas: reuse organization_base as part of package_base

risicle committed:
schemas: collect commonly reused patterns in "common" schema

risicle committed:
schemas: refactor common toplevel response properties into an action_response_base schema

risicle committed:
add schema validation for package_search responses

risicle committed:
schemas: add detail to packages' "tags" and "resources" sections

kentsanggds reviewed (commented):
I wonder if it would be possible to add some tests to test the tests? if its not already covered by a card
Because then I can see what you might expect the response data to look like and tweak the test data to make it fail.
This might also speed up development as you are not dependent on an external source.

risicle commented:
Yeah that's probably not as insane as it sounds.

kentsanggds commented:
Yes, by tests I actually mean add in test data to throw against the E2E tests.

It's a bit meta but it should make it easier to see if the what we are testing against is true against real data and make it easier for a reviewer to test things out.

risicle committed:
add rfc3339-validator and rfc3986-validator dependencies

these will allow us to enforce `format` jsonschema constraints properly

risicle committed:
schemas: enforce format constraints

this includes a hack to replace the default is_datetime checker with a
slightly more lenient version. CKAN's timestamps don't otherwise pass
as a "date-time" because they lack timezone information

risicle committed:
add initial self-tests for schemas against some example responses

risicle commented:
Ok I've now integrated the schema self-tests branch with the beginnings of some of these checks.

kentsanggds commented on ckanfunctionaltests/api/test_schemas.py:
```diff
@@ -0,0 +1,123 @@
+def test_organization_show_is_organization():
+    example_response = _get_example_response("organization_show.json")
+
+    # note non-existent leap-day
```
This comment is probably wrong

kentsanggds replied:
Looks like the comment was removed

risicle committed:
schemas: validate uuids

we have to supply our own format checker for now as this is not yet
supperted by jsonschema

kentsanggds reviewed (changes requested):
Noticed a comment that needs updating but otherwise looks good.

kentsanggds reviewed (approved):
👍
